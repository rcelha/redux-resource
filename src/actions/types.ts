import { DataType } from '../types';

export interface ServiceReturn {
  data: DataType | [DataType];
  meta?: object;
}

export type ServiceFunction = (
  serviceParameters: ServiceOptions
) => Promise<ServiceReturn>;

export interface ResourceOptions {
  append?: boolean;
  ignoreCache?: boolean;
  cached?: boolean;
}

export interface ServiceOptions {
  id?: string | number;
  [propName: string]: any; // tslint:disable-line:no-any
}

export interface ResourceAction {
  type: string;
  resourceType: string;
  payload: {
    serviceFunction: ServiceFunction;
    serviceParameters: ServiceOptions;
    resourceOptions: ResourceOptions;
  };
}

export interface ResourceSuccessAction {
  type: string;
  resourceType: string;
  payload: {
    data: DataType | [DataType];
    meta?: object;
    serviceParameters: ServiceOptions;
    resourceOptions: ResourceOptions;
  };
}

export interface ResourceFailureAction {
  type: string;
  resourceType: string;
  payload: {
    error: Error;
    serviceParameters: ServiceOptions;
    resourceOptions: ResourceOptions;
  };
}
