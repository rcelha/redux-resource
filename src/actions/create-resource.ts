import * as actionTypes from '../action-types';
import { isObject } from 'lodash';
import {
  ServiceFunction,
  ServiceOptions,
  ResourceOptions,
  ResourceAction,
  ServiceReturn,
  ResourceSuccessAction,
  ResourceFailureAction,
} from './types';

/**
 * Triggers the beginning of the resource creation flow
 */
export function createResource(
  resourceType: string,
  serviceFunction: ServiceFunction,
  serviceParameters: ServiceOptions,
  resourceOptions?: ResourceOptions
): ResourceAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_ADD,
    resourceType,
    payload: {
      serviceFunction,
      serviceParameters,
      resourceOptions,
    },
  };
}

/**
 * @ignore
 */
export function createResourceSuccess(
  resourceType: string,
  serviceResult: ServiceReturn,
  serviceParameters: ServiceOptions,
  resourceOptions?: ResourceOptions
): ResourceSuccessAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_ADD_SUCCESS,
    resourceType,
    payload: {
      data: serviceResult.data,
      meta: serviceResult.meta,
      serviceParameters,
      resourceOptions,
    },
  };
}

/**
 * @ignore
 */
export function createResourceFailure(
  resourceType: string,
  error: Error,
  serviceParameters: ServiceOptions,
  resourceOptions?: ResourceOptions
): ResourceFailureAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_ADD_FAILURE,
    resourceType,
    payload: {
      error,
      serviceParameters,
      resourceOptions,
    },
  };
}
