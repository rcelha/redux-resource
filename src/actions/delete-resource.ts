import * as actionTypes from '../action-types';
import {
  ServiceFunction,
  ServiceOptions,
  ResourceOptions,
  ResourceAction,
  ServiceReturn,
  ResourceSuccessAction,
  ResourceFailureAction,
} from './types';

/**
 * Triggers the beginning of the resource deletion flow
 */
export function deleteResource(
  resourceType: string,
  serviceFunction: ServiceFunction,
  serviceParameters: ServiceOptions,
  resourceOptions?: ResourceOptions
): ResourceAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_DEL,
    resourceType,
    payload: {
      serviceFunction,
      serviceParameters,
      resourceOptions,
    },
  };
}

/**
 * @ignore
 */
export function deleteResourceSuccess(
  resourceType: string,
  serviceResult: ServiceReturn,
  serviceParameters: ServiceOptions,
  resourceOptions?: ResourceOptions
): ResourceSuccessAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_DEL_SUCCESS,
    resourceType,
    payload: {
      data: serviceResult.data,
      meta: serviceResult.meta,
      serviceParameters,
      resourceOptions,
    },
  };
}

/**
 * @ignore
 */
export function deleteResourceFailure(
  resourceType: string,
  error: Error,
  serviceParameters: ServiceOptions,
  resourceOptions?: ResourceOptions
): ResourceFailureAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_DEL_FAILURE,
    resourceType,
    payload: {
      error,
      serviceParameters,
      resourceOptions,
    },
  };
}
