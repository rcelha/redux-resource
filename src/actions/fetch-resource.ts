import * as actionTypes from '../action-types';
import { isObject } from 'lodash';
import {
  ServiceFunction,
  ServiceOptions,
  ResourceOptions,
  ResourceAction,
  ServiceReturn,
  ResourceSuccessAction,
  ResourceFailureAction,
} from './types';

/**
 * Triggers the beginning of the resource fetching flow
 */
export function fetchResource(
  resourceType: string,
  serviceFunction: ServiceFunction,
  serviceParameters: string | number | ServiceOptions,
  resourceOptions?: ResourceOptions
): ResourceAction {
  const serviceParametersObject = isObject(serviceParameters)
    ? serviceParameters
    : { id: serviceParameters };
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_GET,
    resourceType,
    payload: {
      serviceFunction,
      serviceParameters: serviceParametersObject,
      resourceOptions,
    },
  };
}

/**
 * @ignore
 */
export function fetchResourceSuccess(
  resourceType: string,
  serviceResult: ServiceReturn,
  serviceParameters: ServiceOptions,
  resourceOptions?: ResourceOptions
): ResourceSuccessAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_GET_SUCCESS,
    resourceType,
    payload: {
      data: serviceResult.data,
      meta: serviceResult.meta,
      serviceParameters,
      resourceOptions,
    },
  };
}

/**
 * @ignore
 */
export function fetchResourceFailure(
  resourceType: string,
  error: Error,
  serviceParameters: ServiceOptions,
  resourceOptions?: ResourceOptions
): ResourceFailureAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_GET_FAILURE,
    resourceType,
    payload: {
      error,
      serviceParameters,
      resourceOptions,
    },
  };
}
