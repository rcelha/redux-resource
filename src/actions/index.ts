export * from './fetch-resource';
export * from './list-resources';
export * from './create-resource';
export * from './delete-resource';
