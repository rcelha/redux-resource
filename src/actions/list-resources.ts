import * as actionTypes from '../action-types';
import {
  ServiceFunction,
  ServiceOptions,
  ResourceOptions,
  ResourceAction,
  ServiceReturn,
  ResourceSuccessAction,
  ResourceFailureAction,
} from './types';

export function listResources(
  resourceType: string,
  serviceFunction: ServiceFunction,
  serviceParameters: ServiceOptions = {},
  resourceOptions?: ResourceOptions
): ResourceAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_LIST,
    resourceType,
    payload: {
      serviceFunction,
      serviceParameters,
      resourceOptions,
    },
  };
}

/**
 * @ignore
 */
export function listResourcesSuccess(
  resourceType: string,
  serviceResult: ServiceReturn,
  serviceParameters: ServiceOptions = {},
  resourceOptions?: ResourceOptions
): ResourceSuccessAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_LIST_SUCCESS,
    resourceType,
    payload: {
      data: serviceResult.data,
      meta: serviceResult.meta,
      serviceParameters,
      resourceOptions,
    },
  };
}

/**
 * @ignore
 */
export function listResourcesFailure(
  resourceType: string,
  error: Error,
  serviceParameters: ServiceOptions = {},
  resourceOptions?: ResourceOptions
): ResourceFailureAction {
  resourceOptions = resourceOptions || {};
  return {
    type: actionTypes.RESOURCE_LIST_FAILURE,
    resourceType,
    payload: {
      error,
      serviceParameters,
      resourceOptions,
    },
  };
}
