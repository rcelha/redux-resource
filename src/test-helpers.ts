import { applyMiddleware, createStore } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { keyBy } from 'lodash';
import { saga } from './index';
import { reducer } from './reducers';

export const sleep = (milliseconds: number) => {
  return new Promise(resolve => setTimeout(resolve, milliseconds));
};

export const setupStore = () => {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(reducer, applyMiddleware(sagaMiddleware));
  sagaMiddleware.run(saga);
  return store;
};

export const service = {
  fetchUser({ id }: { id: number }) {
    const users = service.fetchUsers();
    const data = keyBy(users.data, 'id')[id];
    return { data };
  },

  fetchUsers({ p = 0 } = {}) {
    const data = [
      [{ id: 1, name: 'Rodrigo' }, { id: 2, name: 'Fernanda' }],
      [{ id: 11, name: 'Cherry' }, { id: 12, name: 'Tyrion' }],
      [{ id: 21, name: 'Liz' }, { id: 22, name: 'Edo' }],
    ];

    return {
      data: data[p],
      meta: { total: 2 },
    };
  },

  postUser({ expand = false } = {}) {
    const data = expand
      ? { id: 100, name: 'New User From Service' }
      : { id: 999 };
    return { data };
  },
};
