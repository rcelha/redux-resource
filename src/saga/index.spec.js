import * as actions from '../actions';
import { sleep, setupStore, service } from '../test-helpers';

describe('sagas', () => {
  let store;
  beforeEach(() => {
    store = setupStore();
  });

  describe('when a fetchResource action is dispatched', () => {
    it('executes the service and dispatches a fetchResourceSuccess', async () => {
      const action = actions.fetchResource('users', service.fetchUser, 1);
      store.dispatch(action);
      await sleep(0);
      expect(store.getState().byId.users['1']).toMatchObject({
        data: { id: 1, name: 'Rodrigo' },
        loading: false,
        initialized: true,
      });
    });
  });

  describe('when a listResources action is dispatched', () => {
    it('executes the service and dispatches a listResourcesSuccess', async () => {
      const action = actions.listResources('users', service.fetchUsers);
      store.dispatch(action);
      await sleep(0);
      expect(store.getState().list.users).toMatchObject({
        data: [{ id: 1, name: 'Rodrigo' }, { id: 2, name: 'Fernanda' }],
        loading: false,
        initialized: true,
        error: null,
      });
    });
  });

  describe('when a listResourcesSuccess action is dispatched', () => {
    it('saves its result in `byId` as cached', async () => {
      const action = actions.listResources('users', service.fetchUsers);
      store.dispatch(action);
      await sleep(0);
      expect(store.getState().byId.users).toMatchObject({
        '1': {
          data: { id: 1, name: 'Rodrigo' },
          loading: false,
          cached: true,
        },
        '2': {
          data: { id: 2, name: 'Fernanda' },
          loading: false,
          cached: true,
        },
      });
    });
  });

  describe('when a createResources action is dispatched', () => {
    it('executes the service and dispatches a createResourceSuccess', async () => {
      const action = actions.createResource('users', service.postUser, {
        name: 'Cherry',
      });
      store.dispatch(action);

      await sleep(0);
      expect(store.getState().mutation.users).toMatchObject({
        data: {
          id: 999,
          name: 'Cherry',
        },
      });
    });

    it('saves result to cache', async () => {
      const action = actions.createResource('users', service.postUser, {
        name: 'Cherry',
      });
      store.dispatch(action);

      await sleep(0);
      expect(store.getState().byId.users).toMatchObject({
        '999': {
          data: { id: 999, name: 'Cherry' },
        },
      });
    });
  });

  describe('queries', () => {
    it('has the different queries', async () => {
      let action = actions.listResources('users', service.fetchUsers);
      store.dispatch(action);
      await sleep(0);
      expect(store.getState().query).toMatchSnapshot();

      action = actions.listResources('users', service.fetchUsers, { p: 1 });
      store.dispatch(action);
      await sleep(0);
      expect(store.getState().query).toMatchSnapshot();

      action = actions.listResources('users', service.fetchUsers, { p: 2 });
      store.dispatch(action);
      await sleep(0);
      expect(store.getState().query).toMatchSnapshot();
    });
  });
});
