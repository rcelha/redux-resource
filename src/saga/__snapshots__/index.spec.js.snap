// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sagas queries has the different queries 1`] = `
Object {
  "users": Object {
    "{}": Object {
      "cached": false,
      "data": Array [
        Object {
          "id": 1,
          "name": "Rodrigo",
        },
        Object {
          "id": 2,
          "name": "Fernanda",
        },
      ],
      "error": null,
      "initialized": true,
      "loading": false,
      "meta": Object {
        "total": 2,
      },
    },
  },
}
`;

exports[`sagas queries has the different queries 2`] = `
Object {
  "users": Object {
    "{\\"p\\":1}": Object {
      "cached": false,
      "data": Array [
        Object {
          "id": 11,
          "name": "Cherry",
        },
        Object {
          "id": 12,
          "name": "Tyrion",
        },
      ],
      "error": null,
      "initialized": true,
      "loading": false,
      "meta": Object {
        "total": 2,
      },
    },
    "{}": Object {
      "cached": false,
      "data": Array [
        Object {
          "id": 1,
          "name": "Rodrigo",
        },
        Object {
          "id": 2,
          "name": "Fernanda",
        },
      ],
      "error": null,
      "initialized": true,
      "loading": false,
      "meta": Object {
        "total": 2,
      },
    },
  },
}
`;

exports[`sagas queries has the different queries 3`] = `
Object {
  "users": Object {
    "{\\"p\\":1}": Object {
      "cached": false,
      "data": Array [
        Object {
          "id": 11,
          "name": "Cherry",
        },
        Object {
          "id": 12,
          "name": "Tyrion",
        },
      ],
      "error": null,
      "initialized": true,
      "loading": false,
      "meta": Object {
        "total": 2,
      },
    },
    "{\\"p\\":2}": Object {
      "cached": false,
      "data": Array [
        Object {
          "id": 21,
          "name": "Liz",
        },
        Object {
          "id": 22,
          "name": "Edo",
        },
      ],
      "error": null,
      "initialized": true,
      "loading": false,
      "meta": Object {
        "total": 2,
      },
    },
    "{}": Object {
      "cached": false,
      "data": Array [
        Object {
          "id": 1,
          "name": "Rodrigo",
        },
        Object {
          "id": 2,
          "name": "Fernanda",
        },
      ],
      "error": null,
      "initialized": true,
      "loading": false,
      "meta": Object {
        "total": 2,
      },
    },
  },
}
`;
